core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java

```diff
@@ -2278,9 +2278,12 @@ public Expression implement(
 
       final Type returnType =
           translator.typeFactory.getJavaClass(call.getType());
-      return Types.castIfNecessary(returnType,
-          Expressions.makeBinary(expressionType, expressions.get(0),
-              expressions.get(1)));
+      final Expression result = Expressions.makeBinary(expressionType,
+          expressions.get(0), expressions.get(1));
+      if (returnType == BigDecimal.class) {
+        return RexToLixTranslator.convert(result, returnType);
+      }
+      return Types.castIfNecessary(returnType, result);
     }
 
```

Comment:
Can we move this `BigDecimal` cast logic into `Types.castIfNecessary` ?

Reply:
@danny0405 To the best of my knowledge, data type conversion are implemented in many places (`Types.castIfNecessary`, `RexToLixTranslator.convert`, `RexToLixTranslator.translateCast`, etc.) with some overlaps. We can move this logic into `Types.castIfNecessary`, but it will bring more redundancy, because the logic has been implemented in `RexToLixTranslator.convert` already. 

Therefore, this PR handle it in a case-by-case way. I ever tried to unify these functions and combine them into one. But I do not know whether it is an appropriate way. Maybe @zabetak  , @julianhyde and @vlsi can provide some helpful advices.

Reply:
I don't think `RexToLixTranslator.convert`, `RexToLixTranslator.translateCast` and `Types.castIfNecessary` have any function overlaps, the first 2 are purely about the expression translate, they just produce the expression you want, but `Types.castIfNecessary` contains logic to decide if the is any necessity to do the cast, that is why i suggest you to move the logic.

If you think i'm wrong or other people can do it better, feel free to ignore this.

Reply:
@danny0405 Thanks for discussion and suggestion, Danny. It is great helpful to me to make it more clear.

Current now, I'm a bit of confusing about the semantic differences between `cast `and `convert`. It seems in `Types.castIfNecessary` also conduct some converting work. 
For instance, from **_Integer_** to **_int_**, the cast will call `Expressions.unbox`. Meanwhile, we can also do it with `RexToLixTranslator.convert`. So I am curious about the design intention, but I can't find some documents.

Reply:
`CAST` and `CONVERT` are 2 different sql functions but have the same functionality.

Reply:
Hi, @danny0405, I have updated the PR based on your recent changes.
There are some problems when move the following logic from `RexToLixTranslator.convert` into `Types.castIfNecessary`.
(1) It relies on SqlFunctions, which is impossible to call it in Linq4j.
(2) Make the logic duplicate.
Do you have some suggestions to handle it? Maybe we can unify them?
```
else if (toType == BigDecimal.class) {
      if (fromBox != null) {
        // E.g. from "Integer" to "BigDecimal".
        // Generate "x == null ? null : new BigDecimal(x.intValue())"
        return Expressions.condition(
            Expressions.equal(operand, RexImpTable.NULL_EXPR),
            RexImpTable.NULL_EXPR,
            Expressions.new_(
                BigDecimal.class,
                Expressions.unbox(operand, fromBox)));
      }
      if (fromPrimitive != null) {
        // E.g. from "int" to "BigDecimal".
        // Generate "new BigDecimal(x)"
        // Fix CALCITE-2325, we should decide null here for int type.
        return Expressions.condition(
            Expressions.equal(operand, RexImpTable.NULL_EXPR),
            RexImpTable.NULL_EXPR,
            Expressions.new_(
                BigDecimal.class,
                operand));
      }
      // E.g. from "Object" to "BigDecimal".
      // Generate "x == null ? null : SqlFunctions.toBigDecimal(x)"
      return Expressions.condition(
          Expressions.equal(operand, RexImpTable.NULL_EXPR),
          RexImpTable.NULL_EXPR,
          Expressions.call(
              SqlFunctions.class,
              "toBigDecimal",
              operand));
    }
```

Reply:
I kind of think we should invoke `RexToLixTranslator#convert` in the last step of the `Types#castIfNecessary` instead of the `Expressions.convert_`[1], what fo you think ?

[1] https://github.com/apache/calcite/blob/955d4ea7b30b85519529c1c77662fa04141bfa89/linq4j/src/main/java/org/apache/calcite/linq4j/tree/Types.java#L464

Reply:
Yes, I agree. In my impression, there may be some tricks need to be handled. I will give a try. Thanks!